cmd/notation/inspect.go

```diff
@@ -0,0 +1,288 @@
+	treeItemPrefixLast = "└──"
+	subTreePrefix      = "│  "
+	subTreePrefixLast  = "   "
+)
+
+type inspectOpts struct {
+	cmd.LoggingFlagOpts
+	SecureFlagOpts
+	reference    string
+	outputFormat string
+}
+
+type inspectOutput struct {
+	MediaType  string `json:"mediaType"`
+	Signatures []signatureOutput
+}
+
+type signatureOutput struct {
+	Digest                string              `json:"digest"`
+	SignatureAlgorithm    string              `json:"signatureAlgorithm"`
+	SignedAttributes      map[string]string   `json:"signedAttributes"`
+	UserDefinedAttributes map[string]string   `json:"userDefinedAttributes"`
+	UnsignedAttributes    map[string]string   `json:"unsignedAttributes"`
+	Certificates          []certificateOutput `json:"certificates"`
+	SignedArtifact        artifact            `json:"signedArtifact"`
+}
+
+type certificateOutput struct {
+	SHA1Fingerprint string `json:"SHA1Fingerprint"`
+	IssuedTo        string `json:"issuedTo"`
+	IssuedBy        string `json:"issuedBy"`
+	Expiry          string `json:"expiry"`
+}
+
+type artifact struct {
+	MediaType string `json:"mediaType"`
+	Digest    string `json:"digest"`
+	Size      int64  `json:"size"`
+}
+
+func inspectCommand(opts *inspectOpts) *cobra.Command {
+	if opts == nil {
+		opts = &inspectOpts{}
+	}
+	command := &cobra.Command{
+		Use:   "inspect [reference]",
+		Short: "Inspect all signatures associated with the signed artifact",
+		Long:  "Inspect all signatures associated with the signed artifact.",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("missing reference")
+			}
+			opts.reference = args[0]
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runInspect(cmd, opts)
+		},
+	}
+
+	opts.LoggingFlagOpts.ApplyFlags(command.Flags())
+	opts.SecureFlagOpts.ApplyFlags(command.Flags())
+	cmd.SetPflagOutput(command.Flags(), &opts.outputFormat, cmd.PflagOutputUsage)
+	return command
+}
+
+func runInspect(command *cobra.Command, opts *inspectOpts) error {
+	// set log level
+	ctx := opts.LoggingFlagOpts.SetLoggerLevel(command.Context())
+
+	if opts.outputFormat != ioutil.OutputJson && opts.outputFormat != ioutil.OutputPlaintext {
+		return fmt.Errorf("unrecognized output format %s", opts.outputFormat)
+	}
+
+	// initialize
+	reference := opts.reference
+	sigRepo, err := getSignatureRepository(ctx, &opts.SecureFlagOpts, reference)
+	if err != nil {
+		return err
+	}
+
+	manifestDesc, ref, err := getManifestDescriptor(ctx, &opts.SecureFlagOpts, reference, sigRepo)
+	if err != nil {
+		return err
+	}
+
+	// reference is a digest reference
+	if err := ref.ValidateReferenceAsDigest(); err != nil {
+		if opts.outputFormat == ioutil.OutputPlaintext {
+			fmt.Printf("Resolved artifact tag `%s` to digest `%s` before inspect.\n", ref.Reference, manifestDesc.Digest.String())
+			fmt.Println("Warning: The resolved digest may not point to the same signed artifact, since tags are mutable.")
```

Comment:
I'm trying to understand this printout, since `ref` is a digest reference (validated by `ref.ValidateReferenceAsDigest()`), we are not resolving any tag here right?

Reply:
```
	// reference is a digest reference
	if err := ref.ValidateReferenceAsDigest(); err != nil {
		if opts.outputFormat == ioutil.OutputPlaintext {
			fmt.Printf("Resolved artifact tag `%s` to digest `%s` before inspect.\n", ref.Reference, manifestDesc.Digest.String())
			fmt.Println("Warning: The resolved digest may not point to the same signed artifact, since tags are mutable.")
		}
		ref.Reference = manifestDesc.Digest.String()
	}
```

the if condition is different from the one in resolveReference. the warning is printed if the err from ref.ValidateReferenceAsDigest() != nil (so the reference is a tag)